### PR TITLE
add EXP_DUAL to supported cones

### PR DESCRIPTION
**Pull request by sbarratt**

### CODE REVIEW

File: cpp/src/cones.cpp [hunk 1]
```diff
@@ -195,7 +195,9 @@ LinearOperator _dprojection_exp(const Vector &x, bool dual) {
       double t = 0;
       double rs[3] = {x_i[0], x_i[1], x_i[2]};
       int ret = _proj_exp_cone(rs, &t);
-      assert(ret == 0);
+      if (ret != 0) {
+        throw "Projection onto exponential cone failed.";
```
Comment (akshayka) [inline cpp/src/cones.cpp]: throw a runtime_error? https://docs.microsoft.com/en-us/cpp/standard-library/stdexcept?view=vs-2019

File: cpp/src/cones.cpp [hunk 2]
```diff
@@ -324,9 +326,12 @@ LinearOperator _dprojection(const Vector &x, ConeType type, bool dual) {
...
   } else if (type == PSD) {
     return _dprojection_psd(x);
-  } else {
-    assert(type == EXP);
+  } else if (type == EXP) {
     return _dprojection_exp(x, dual);
+  } else if (type == EXP_DUAL) {
+    return _dprojection_exp(x, !dual);
+  } else {
+    throw "Cone not supported.";
```
Comment (akshayka) [inline cpp/src/cones.cpp]: `throw invalid_argument("Cone not supported")`? https://docs.microsoft.com/en-us/cpp/standard-library/stdexcept?view=vs-2019

File: cpp/src/cones.cpp [hunk 3]
```diff
@@ -407,9 +416,14 @@ void _dprojection_dense(MatrixRef &D_block, const Vector &x, ConeType type,
...
-  } else {
-    assert(type == EXP);
+  } else if (type == EXP) {
+    // TODO: Should be able to do this with one solve
     _op_into_dense(D_block, _dprojection_exp(x, dual));
+  } else if (type == EXP_DUAL) {
+    // TODO: Should be able to do this with one solve
+    _op_into_dense(D_block, _dprojection_exp(x, !dual));
+  } else {
+    throw "Cone not supported.";
```
Comment (akshayka) [inline cpp/src/cones.cpp]: throw invalid_argument("Cone not supported")

File: diffcp/cones.py [hunk 4]
```diff
@@ -132,7 +136,7 @@ def pi(x, cones, dual=False):
         for dim in sz:
             if cone == PSD:
                 dim = vec_psd_dim(dim)
-            elif cone == EXP:
+            elif cone == EXP or cone == EXP_DUAL:
```
Comment (akshayka) [inline diffcp/cones.py]: The second check (`cone == EXP_DUAL`) is unnecessary, since you overwrite cone above

Comment (sbarratt) [reply diffcp/cones.py]: I actually need this, since I change EXP_DUAL to EXP inside `_proj`.

Comment (akshayka) [reply diffcp/cones.py]: op sorry, github collapsed all the code in between and I thought that lines 82-85 were a part of this function.

**Comment by akshayka** [comment]
lgtm

**Pull request merged**